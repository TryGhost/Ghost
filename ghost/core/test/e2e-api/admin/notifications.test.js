const should = require('should');
const {agentProvider, fixtureManager, matchers} = require('../../utils/e2e-framework');
const {anyContentVersion, anyObjectId, anyEtag, anyLocationFor} = matchers;

const matchNotification = {
    id: anyObjectId
};

describe('Notifications API', function () {
    let agent;

    before(async function () {
        agent = await agentProvider.getAdminAPIAgent();
        await fixtureManager.init();
        await agent.loginAsOwner();
    });

    it('Can add notification', async function () {
        const newNotification = {
            type: 'info',
            message: 'test notification',
            custom: true,
            id: '59a952be7d79ed06b0d21133'
        };

        await agent
            .post('notifications')
            .body({
                notifications: [newNotification]
            })
            .expectStatus(201)
            .matchBodySnapshot()
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            });
    });

    it('Can delete notification', async function () {
        const newNotification = {
            type: 'info',
            message: 'test notification',
            status: 'alert',
            custom: true
        };

        // create the notification to deleted
        const res = await agent
            .post('notifications')
            .body({
                notifications: [newNotification]
            })
            .matchBodySnapshot({
                notifications: [matchNotification]
            })
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag,
                location: anyLocationFor('notifications')
            });

        const jsonResponse = res.body;

        should.exist(jsonResponse.notifications);
        jsonResponse.notifications.length.should.eql(1);

        jsonResponse.notifications[0].type.should.equal(newNotification.type);
        jsonResponse.notifications[0].message.should.equal(newNotification.message);
        jsonResponse.notifications[0].status.should.equal(newNotification.status);

        const notification = jsonResponse.notifications[0];

        await agent
            .delete(`notifications/${notification.id}/`)
            .matchBodySnapshot()
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .expectStatus(204);

        const res3 = await agent
            .get('notifications')
            .matchBodySnapshot({
                notifications: [matchNotification]
            })
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .expectStatus(200);

        const deleted = res3.body.notifications.filter(n => n.id === notification.id);
        deleted.should.be.empty();
    });
});
