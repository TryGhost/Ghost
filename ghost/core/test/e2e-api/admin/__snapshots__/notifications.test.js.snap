// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications API Can add notification 1: [body] 1`] = `
Object {
  "notifications": Array [
    Object {
      "custom": true,
      "dismissible": true,
      "id": "59a952be7d79ed06b0d21133",
      "location": "bottom",
      "message": "test notification",
      "status": "alert",
      "type": "info",
    },
  ],
}
`;

exports[`Notifications API Can add notification 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "167",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "location": "http://127.0.0.1:2369/ghost/api/admin/notifications/59a952be7d79ed06b0d21133/",
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Notifications API Can delete notification 1: [body] 1`] = `
Object {
  "notifications": Array [
    Object {
      "custom": true,
      "dismissible": true,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "location": "bottom",
      "message": "test notification",
      "status": "alert",
      "type": "info",
    },
  ],
}
`;

exports[`Notifications API Can delete notification 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "167",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "location": StringMatching /https\\?:\\\\/\\\\/\\.\\*\\?\\\\/notifications\\\\/\\[a-f0-9\\]\\{24\\}\\\\//,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Notifications API Can delete notification 3: [body] 1`] = `Object {}`;

exports[`Notifications API Can delete notification 4: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin",
  "x-powered-by": "Express",
}
`;

exports[`Notifications API Can delete notification 5: [body] 1`] = `
Object {
  "notifications": Array [
    Object {
      "custom": true,
      "dismissible": true,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "location": "bottom",
      "message": "test notification",
      "status": "alert",
      "type": "info",
    },
  ],
}
`;

exports[`Notifications API Can delete notification 6: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "167",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;
